### PR TITLE
[Config] Add GLM-5.2 SFT config surface and fused-expert export

### PR DESCRIPTION
**Pull request by zrr1999**
#### Before submitting

- [x] Lint code. If there are lint issues, please format the code first.

```shell
# Install and register `pre-commit` in the project folder
pip install pre-commit && pre-commit install

# Process previous code files separately
pre-commit run --file XXXX.py
```

- [x] Add test cases into `tests` folder. If there are codecov issues, please add tests cases first.

### PR types
New features

### PR changes
Models

### Description

This PR is superseded by [#4956](https://github.com/PaddlePaddle/PaddleFormers/pull/4956), which contains the complete GLM-5.2 implementation against develop. This PR is closed without merging; the implementation continues in the linked PR.

First GLM-5.2 structure layer in PaddleFormers: construct and export from published HF `config.json` / SFT YAML. No accuracy kernels.

Third-party stack aligned to HF config and chat template, not an official trainer.

- `template: glm5_2` uses a full empty thought pair (`<think></think>`). `enable_thinking` (default False on `GeneratingArguments`) is copied into `dataset_config` only for `glm5_2`. Copying onto all SFT templates would change Qwen3-VL (`qwen3_vl` default True).
- After `set_llm_config()`, copy YAML/CLI MTP depth, `moe_expert_fusion`, PP P2P (`overlap_p2p_comm` / `batch_p2p_comm`), parallel sizes, and `persist_layer_norm` onto the Fleet provider. `TransformerConfig` rejects setting `mtp_num_layers` directly; P2P lives on `TrainingArguments`. Do not copy `variable_seq_lengths` (YAML only drives pipeline `enable_dynamic_shape`).
- `GlmMoeDsaModelProvider` maps `expert_tensor_model_parallel_size` → `expert_tensor_parallel_size`. Putting this on shared `GLMMoEModelProvider` would break glm4_moe unit tests.
- `GlmMoeDsaConfig` round-trips `indexer_rope_interleave`, `rope_parameters.rope_theta`, and `use_qk_norm=true`; derived keys are unsavable for ConfigTester.
- `save_to_hf` default False; fused grouped-GEMM restore accepts 3-D expert weights instead of aborting on 2-D. Default True would start HF export on existing jobs.
- `SFTDataset` with `packing=false` no longer yields the last sample twice after StopIteration.

Out of scope: glm4_moe weight maps, global tokenizer `enable_thinking=False`, accuracy kernels. This structural Draft is a dependency of the combined PaddleFormers tip, not standalone training.

Fused-expert optimizer saves expose temporary 2-D views only for the corresponding grouped-expert states. Accumulator and FP32 master-weight mappings are restored to their original objects in `finally`; unrelated 3-D states and live tensor shapes are preserved. Real save/load/next-step tests cover FP32, BF16 and failure paths.

Validation used these combined stack heads on H800/CUDA/bfloat16:

- [PaddleFleet `d5a5379f1263`](https://github.com/PaddlePaddle/PaddleFleet/commit/d5a5379f12637e4dcdd0d07e484b194f8ca0c5cf)
- [PaddleFormers `ffde408b986e`](https://github.com/PaddlePaddle/PaddleFormers/commit/ffde408b986ea811269218e861d11ae87897987d)
- [Megatron-LM `d8b684169735`](https://github.com/zrr1999/Megatron-LM/commit/d8b6841697350023f7b71a5792dd636d9cf63023)
- [ms-swift `91a3bb86d72f`](https://github.com/zrr1999/ms-swift/commit/91a3bb86d72fb8e301bb1a920b7815846c02820e)

| Native entrypoint | Configuration | Steps per framework | Result |
|---|---|---:|---|
| Formal training CLI | TP2 / PP2 / EP1 / SP, GBS1 | 100 | Strict loss, provenance and checkpoint checks passed; raw loss and 187 final tensors bitwise identical |
| GLM52 precision case | TP1 / PP2 / EP2, GBS2 | 100 | Downloaded model shards loaded by both frameworks; raw loss and 187 final tensors bitwise identical |

Both runs reproduce the corresponding pre-change loss bits and updated parameters. Formal invocation: `formal-pr-review-20260909-a-upload-n100-r1`; native case: `ci-glm52-upload-20260909-n100-r1`. The native case used the existing local dataset override; it is not a remote CI execution or a claim about the remote Minimax data cache.

Focused validation includes 34 DSA configuration/layout tests and 13 Trainer tests. The latter execute FP32 and BF16 AdamW save → load → next-step comparisons, check original state object/shape restoration after save failures, and exercise the model-config accuracy guard. Relevant repository format hooks passed.

The official-checkpoint subset is available as [PaddleFormers/GLM-5.2-BF16-minimal](https://git.aistudio.baidu.com/PaddleFormers/GLM-5.2-BF16-minimal.git) at `685a4b891d004a64bd2f1fef320deeffe162af02`. Independent remote download verified all 13 files; both weight shards preserve the original 187 tensor names, shapes, dtypes and bytes. Source: `zai-org/GLM-5.2@b4734de4facf877f85769a911abafc5283eab3d9`.

Remote CI still requires its model cache and matching companion builds. Known failures include review approval, a PaddleFleet wheel missing the deferred-token API, absent GLM4 cache metadata, ms-swift runner permissions/upstream formatting, and Minimax/GLM45 accuracy mismatches. Shared workflows and other model cases remain unchanged. Local GLM52 validation does not claim those checks passed.

Runtime remains Paddle `3.4.0.post20260808+733f3454aa0`, Torch `2.12.1+cu129`, cuBLAS `12.9.1.4`. Frozen source pins and the known Paddle wheel metadata dependency conflict remain separate from this implementation validation; canonical sealed acceptance is not claimed.

Current PR head: `9bedea4973716aff6a457418c425be1c9e237bf3`.

### CODE REVIEW

File: paddleformers/trainer/trainer.py [hunk 1]
```diff
@@ -1187,8 +1238,19 @@ def _save_flex_optimizer_state(self, output_dir):
         optimizer_state_dict_path = os.path.join(output_dir, OPTIMIZER_STATE_DIC)
         optimizer_states = {}
         master_weights = {}
+        if getattr(self.args, "moe_expert_fusion", False):
```
[the remark this answers is not in the document]
Comment (zrr1999) [reply paddleformers/trainer/trainer.py]: Done. `0af5d798ae` 起 fused-expert FlexCheckpoint 的 optimizer 保存改为 fail closed：`moe_expert_fusion=True` 时直接 `RuntimeError`，不再静默跳过 optimizer / master weights / 完成信号。原生 `--sharding fsdp` 路径、公共 TrainingArguments 默认值、GLM4 gate dtype 以及 `save_pretrained` 的官方 `rope_parameters` 已一并恢复。

**Comment by zrr1999** [comment]
当前 head `0af5d798ae` 已处理本轮 P1：恢复 develop FSDP / 公共默认值 / GLM4 gate dtype，fused-expert optimizer save fail-closed，`save_pretrained` 保留官方 `rope_parameters`。请 @Paddle-Bot 对当前 head 重新评审。

File: paddleformers/trainer/trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zrr1999) [reply paddleformers/trainer/trainer.py]: Done on `8b23bc1427`：去掉无条件 `RuntimeError` / 提前 `return`。`moe_expert_fusion=True` 时继续走 FlexCheckpoint：先 `restore_fused_expert_3d_layout(..., optimizer=self.optimizer)` 把 grouped-GEMM moment 展成 Flex 可写的 2-D，再写 optimizer / master weights / `saved_signal_<rank>`。H20 `qwen3vl_sft_moe.yaml`（fusion + sharding stage1 + flex_checkpoint）会走到这条路径，不能再 fail-closed。覆盖：`test_optimizer_path_restores_3d_layout_then_saves`、`test_fused_expert_optimizer_round_trip_keeps_3d_moments`。

**Comment by zrr1999** [comment]
当前 head `8dd69abe04` 恢复 develop 的 `AutoProcessor` 加载路径，以及 UAC 下 `max_grad_norm=0` 的 clip-off。此前 text SFT 把 `processor` 指到 tokenizer，GLM4 CI first-train/resume loss 偏离 GT（`12.635027885` vs `12.63612175`）。请 @Paddle-Bot 对当前 head 重新评审。

**Comment by zrr1999** [comment]
@Paddle-Bot 请对当前 head `46c57e5d1a` 重新 review。本提交只把 CI black 22.8.0 / isort 5.11.5 应用到 GLM-5.2 触及文件，语义未改。
